tests/devices/qubit/test_apply_operation.py

```diff
@@ -68,6 +84,133 @@ def matrix(self):
+        qml.MultiControlledX(wires=[0, 1, 2, 3, 4], control_values=[1, 1, 1, 1]),
+        qml.GroverOperator(wires=[0, 1, 2]),
+        qml.IsingXX(np.pi / 2, wires=[0, 1]),
+        qml.DoubleExcitation(np.pi / 4, wires=[0, 1, 2, 3]),
+    ]
+
+    def test_sparse_operation_dense_state(self):
+        """Test that apply_operation works with a sparse matrix operation"""
+
+        # Create a random unitary matrix
+        U = unitary_group.rvs(2**3)
+        U = qml.math.asarray(U, like="numpy")
+
+        # Create a random state vector
+        state = np.random.rand(2**3) + 1j * np.random.rand(2**3)
+        state = qml.math.asarray(state, like="numpy").reshape([2] * 3)
+
+        # Apply the operation
+        U_sp = qml.QubitUnitary(csr_matrix(U), wires=range(3))
+        new_state = apply_operation_csr_matrix(U_sp, state)
+        expected_state = state.reshape((1, 8)) @ U.reshape((8, 8)).T
+        expected_state = expected_state.reshape([2] * 3)
+
+        assert qml.math.allclose(new_state, expected_state)
+
+    def test_sparse_operation_sparse_state(self):
+        """Test that apply_operation does not support with a sparse state operation"""
+
+        # Create a random unitary matrix
+        U = unitary_group.rvs(2**3)
+        U = qml.math.asarray(U, like="numpy")
+
+        # Create a random state vector
+        state = np.random.rand(2**3) + 1j * np.random.rand(2**3)
+        state = csr_matrix(state)
+        U_sp = qml.QubitUnitary(csr_matrix(U), wires=range(3))
+
+        # Apply the operation
+        with pytest.raises(
+            TypeError,
+            match="State should not be sparse",
+        ):
+            apply_operation_csr_matrix(U_sp, state)
+
+    @pytest.mark.parametrize("N", range(2, 20, 2))
+    def test_sparse_operation_large_N(self, N):
+        """Test that apply_operation_csr_matrix works with a large number of wires"""
+        # Make a sparse unitary matrix by tensor producting several smaller unitaries
+
+        U_list = [I, X, Y, Z]
+        U_list = [csr_matrix(op) for op in U_list]
+        # Pick random indices, then choose unitaries
+        indices = np.random.choice(len(U_list), size=N)
+        Us = [U_list[idx] for idx in indices]
+
+        U = Us[0]
+        for i in range(1, N):
+            U = kron(U, Us[i], format="csr")
+
+        state_shape = (2,) * N
+        state_size = 2**N
+        state = np.random.rand(state_size) + 1j * np.random.rand(state_size)
+        state = state / np.linalg.norm(state)
+        state = state.reshape(state_shape)
+
+        U_sp = qml.QubitUnitary(csr_matrix(U), wires=range(N))
+        new_state = apply_operation_csr_matrix(U_sp, state)
+
+        # Don't waste time constructing dense U to test, instead we just check that the U^Dagger @ state is correct
+        final_state = apply_operation_csr_matrix(U_sp, new_state)
+        assert qml.math.allclose(final_state, state)
+
+    @pytest.mark.parametrize("N", range(2, 20, 2))
```

Comment:
Do we need to run this for such large systems? The previous test checks that things to suddenly a new logical structure at large systems, so I imagine smaller systems should be sufficient.

Reply:
OIC. I'll downsize them. I chose them since initially expected sparse pipeline to be work well between 20 qubits and 30 qubits. But for CI bundles I agree that this is not necessary.

Reply:
https://github.com/PennyLaneAI/pennylane/pull/6883/commits/f157d9177379b681645d42b96475ec66c6694a59